debian/anonsurf.1
.TH AnonSurf
.SH Name
anonsurf \- Anonymizaton toolkit for Parrot Security OS
.SH Usage
AnonSurf [start | stop | restart | changeid | status | myip | dns]
.SH DESCRIPTION
.PP
Setup Tor proxy to route all traffic through Tor

* Modules:
  * UX: 
    * CLI (anonsurf): Handle user's options from command line interface.
    * GUI (anonsurf-gtk): A new friendly GTK GUI
  * Daemon: By now AnonSurf supports systemd only
    * anondaemon: A bash script to setup Tor transparent proxy
    * make_torrc: a simple script to handle configurations of session
  * Extra tool:
    * dnstool: Quick cli tool to change system DNS setting

* Features:
  * Start / stop / restart AnonSurf
  * Check current Tor status using Nyx
  * Change your ID by restarting to switch Tor nodes
  * Check your current public IP
  * Check if AnonSurf is enabled / disbled at boot
  * Enable / disable AnonSurf at boot (as system service)

* start:
  * Use bleachbit to kill dangerous applications (chrome skype thunderbird firefox...) 
  * Create system daemon starting Tor proxy gate
  * Disable IPv6 (security/IPv6 may reveal permanent MAC address)
  * Routes all internet packets/traffic through Tor
  * All DNS through Tor

* changeid:
  * Changes Tor node giving us new IP address using control port

* status:
  * Use Nyx application to display information about Tor service, bandwith, nodes...

* myip:
  * Check current IP address by checking response from Tor server

* status-boot: Show if AnonSurf is enabled / disabled
* enable-boot: Use systemctl (from systemd or 3rd package) to enable at boot
* disable-boot: Use systemctl (from systemd or 3rd package) to disable at boot

* dnstool:
  * Check current status: Dynamic / Static setting.
    * Dynamic: System use DCHP's settings
    * Static: Do not use DHCP's settings. Custom settings won't change after reboot
  * Set system DNS with Dynamic:
    * Use only DHCP server
    * Add more servers in tail (requires resolvconf). Extra servers won't change after reboot
  * Set Static DNS:
    * Repace DNS in /etc/resolv.conf
  * This tool supports OpenNIC servers by default.

.PP


.SH AUTHOR
Developed by:
 * Lorenzo "Palinuro" Faletra <palinuro@parrotsec.org>
 * Lisetta "Sheireen" Ferrero <sheireen@parrotsec.org>
 * Francesco "Mibofra" Bonanno <mibofra@parrotsec.org>

Maintained by:
 * Nong Hoang "DmKnght" Tu <dmknght@parrotsec.org>

.PP
This manual and package was written by Nong Hoang "DmKnght" Tu <dmknght@parrotsec.org> and RightToPrivacy <righttoprivacy@tutanota.com>


debian/changelog
anonsurf (3.0.5+parrot0) rolling-testing; urgency=medium

  * Daemon: Fix missing reset color causes cli issue
  * GUI: Optimized ram usage
  * GUI: Use embedded icons instead of theme icons
  * GUI: Now use native start / stop / restart functions
  * CLI: completely new CLI in Nim lang
  * Fix bleachbit command (invalid module) and improve clean list

 -- Nong Hoang "DmKnght" Tu <dmknght@parrotsec.org>  Wed, 16 Sep 2020 13:56:50 +0700

anonsurf (3.0.4+parrot0) rolling-testing; urgency=medium

  * Disable resize in GUI
  * Fix icon path for new parrot-menu update
  * Better color display for dnstool

 -- Nong Hoang "DmKnght" Tu <dmknght@parrotsec.org>  Sat, 05 Sep 2020 14:29:58 +0700

anonsurf (3.0.3+parrot2) rolling-testing; urgency=medium

  * Sorry guys (fixing typo).

 -- Lorenzo "Palinuro" Faletra <palinuro@parrotsec.org>  Fri, 14 Aug 2020 18:23:37 +0200

anonsurf (3.0.3+parrot1) rolling-testig; urgency=medium

  * Fix syntax error in anondaemon.
  * Better handle logged users detection.

 -- Lorenzo "Palinuro" Faletra <palinuro@parrotsec.org>  Thu, 13 Aug 2020 00:31:50 +0200

anonsurf (3.0.2+parrot0) rolling-testing; urgency=medium

  * Add bridge mode
  * Add new firefox name to kill
  * New GUI layout

 -- Nong Hoang "dmknght" Tu <dmknght@parrotsec.org>  Fri, 31 Jul 2020 04:46:48 +0700

anonsurf (3.0.1+parrot0) rolling-testing; urgency=medium

  * Improve output of dnstool
  * Fix spacing in GUI when showing deactivated ports
  * New icon for boot status

 -- Nong Hoang "dmknght" Tu <dmknght@parrotsec.org>  Mon, 27 Jul 2020 15:14:54 +0700

anonsurf (3.0.0) rolling-testing; urgency=medium

  * Promote previous version to 3.0.0.

 -- Lorenzo "Palinuro" Faletra <palinuro@parrotsec.org>  Tue, 21 Jul 2020 16:07:01 +0200

anonsurf (2.14.0) rolling-testing; urgency=medium

  * Add test version for GTK GUI
  * Add notification when start / stop anonsurf failed
  * Add new icons
  * Mirror improvements
  * Bump development version

 -- Nong Hoang "DmKnght" Tu <dmknght@parrotsec.org>  Mon, 20 Jul 2020 00:00:22 +0700

anonsurf (2.13.9+parrot2) rolling-testing; urgency=medium

  * Remove start-bridge from completion
  * Update nim minimum version and license for nim src

 -- Nong Hoang "DmKnght" Tu <dmknght@parrotsec.org>  Sat, 11 Jul 2020 09:53:02 +0700

anonsurf (2.13.9+parrot1) rolling-testing; urgency=medium

  * Change the logic of restart and start tor

 -- Nong Hoang "DmKnght" Tu <dmknght@parrotsec.org>  Sun, 05 Jul 2020 13:46:42 +0700

anonsurf (2.13.9+parrot0) rolling-testing; urgency=medium

  * Disable bridge support
  * Edit color, typo in output message
  * Fix a bug in anonsurf daemon that use old make-torrc path

 -- Nong Hoang "DmKnght" Tu <dmknght@parrotsec.org>  Sun, 05 Jul 2020 08:25:36 +0700

anonsurf (2.13.8+parrot3) rolling-testing; urgency=medium

  * Typo and code comment fix

 -- Nong Hoang "DmKnght" Tu <dmknght@parrotsec.org>  Thu, 02 Jul 2020 13:54:31 +0700

anonsurf (2.13.8+parrot2) rolling-testing; urgency=medium

  * Move make-torrc to /usr/share/
  * Fix some litian warning
  * Code clean

 -- Nong Hoang "DmKnght" Tu <dmknght@parrotsec.org>  Thu, 02 Jul 2020 06:17:14 +0700

anonsurf (2.13.8+parrot1) rolling-testing; urgency=medium

  * Fix a bug in password generation
  * Rework changeid function
  * Show network error if check IP return empty

 -- Nong Hoang "DmKnght" Tu <dmknght@parrotsec.org>  Thu, 02 Jul 2020 05:28:14 +0700

anonsurf (2.13.7+parrot0) rolling-testing; urgency=medium

  * Add new check IP function with tor project service
  * Remove postinit script from debian packaging
  * Disable systemd auto enable during installation
  * Add random password generation (buggy)
  * Status command now open nyx with custom nyxrc (w/ passwd)

 -- Nong Hoang "DmKnght" Tu <dmknght@parrotsec.org>  Tue, 30 Jun 2020 14:18:15 +0700

anonsurf (2.13.6+parrot0) rolling-testing; urgency=medium

  * Handle missing resolv.conf file
  * make-torrc now create torrc directly
  * Fix a bug in restore backup of DNS settings

 -- Nong Hoang "DmKnght" Tu <dmknght@parrotsec.org>  Mon, 29 Jun 2020 16:22:16 +0700

anonsurf (2.13.5+parrot2) rolling-testing; urgency=medium

  * Fix a logic bug in check valid commands
  * Fix the way to call create-backup

 -- Nong Hoang "DmKnght" Tu <dmknght@parrotsec.org>  Thu, 18 Jun 2020 04:47:09 +0700

anonsurf (2.13.5+parrot1) rolling-testing; urgency=medium

  * Add status-boot to check status of enable at boot
  * Reduce delay time of stop function
  * Try copyFile instead of createSymlink for make-torrc

 -- Nong Hoang "DmKnght" Tu <dmknght@parrotsec.org>  Thu, 18 Jun 2020 04:20:56 +0700

anonsurf (2.13.5+parrot0) rolling-testing; urgency=medium

  * Change the way to check boot status

 -- Nong Hoang "DmKnght" Tu <dmknght@parrotsec.org>  Thu, 18 Jun 2020 03:46:10 +0700

anonsurf (2.13.4+parrot0) rolling-testing; urgency=medium

  * Fix typo in msg
  * More small terminal friendly for help banner
  * Fix empty resolv.conf bug

 -- Nong Hoang "DmKnght" Tu <dmknght@parrotsec.org>  Thu, 18 Jun 2020 03:31:42 +0700

anonsurf (2.13.3+parrot0) rolling-testing; urgency=medium

  * Replace torrc by make-torrc
  * Create test package

 -- Nong Hoang "DmKnght" Tu <dmknght@parrotsec.org>  Fri, 22 May 2020 11:04:27 +0700

anonsurf (2.13.1+parrot0) rolling; urgency=medium

  * Use new code structure for making torrc
  * Use new code structure to start tor service

 -- Nong Hoang "DmKnght" Tu <dmknght@parrotsec.org>  Mon, 04 May 2020 04:35:25 +0700

anonsurf (2.13.0+parrot1) rolling-testing; urgency=medium

  * Correct the configuration for Obfs4Bridge
  * Add "meu" to support list

 -- Nong Hoang "DmKnght" Tu <dmknght@parrotsec.org>  Sun, 03 May 2020 08:33:46 +0700

anonsurf (2.13.0+parrot0) rolling-testing; urgency=medium

  * Add obfs4 bridge support in config generation

 -- Nong Hoang "DmKnght" Tu <dmknght@parrotsec.org>  Sun, 03 May 2020 07:58:45 +0700

anonsurf (2.12.2+parrot0) rolling; urgency=medium

  * Add basic bash completion for anonsurf


anonsurf.sh
#!/bin/bash

### BEGIN INIT INFO
# Provides:          anonsurf
# Required-Start:
# Required-Stop:
# Should-Start:
# Default-Start:
# Default-Stop:
# Short-Description: Transparent Proxy through TOR.
### END INIT INFO
#
# Devs:
# Lorenzo 'Palinuro' Faletra <palinuro@parrotsec.org>
# Lisetta 'Sheireen' Ferrero <sheireen@autistiche.org>
# Francesco 'Mibofra' Bonanno <mibofra@parrotsec.org>
#
# Extended:
# Daniel 'Sawyer' Garcia <dagaba13@gmail.com>
#
# anonsurf is free software: you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
# You can get a copy of the license at www.gnu.org/licenses
#
# anonsurf is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Parrot Security OS. If not, see <http://www.gnu.org/licenses/>.







export BLUE='\033[1;94m'
export GREEN='\033[1;92m'
export RED='\033[1;91m'
export RESETCOLOR='\033[1;00m'


# Destinations you don't want routed through Tor
TOR_EXCLUDE="192.168.0.0/16 172.16.0.0/12 10.0.0.0/8"

# The UID Tor runs as
# change it if, starting tor, the command 'ps -e | grep tor' returns a different UID
TOR_UID="debian-tor"

# Tor's TransPort
TOR_PORT="9040"









function notify {
	if [ -e /usr/bin/notify-send ]; then
		/usr/bin/notify-send "AnonSurf" "$1"
	fi
}
export notify


function init {
	echo -e -n "$BLUE[$GREEN*$BLUE] killing dangerous applications\n"
	sudo killall -q chrome dropbox iceweasel skype icedove thunderbird firefox firefox-esr chromium xchat hexchat transmission steam firejail
	echo -e -n "$BLUE[$GREEN*$BLUE] Dangerous applications killed\n"
	notify "Dangerous applications killed"

	echo -e -n "$BLUE[$GREEN*$BLUE] cleaning some dangerous cache elements\n"
	bleachbit -c adobe_reader.cache chromium.cache chromium.current_session chromium.history elinks.history emesene.cache epiphany.cache firefox.url_history flash.cache flash.cookies google_chrome.cache google_chrome.history  links2.history opera.cache opera.search_history opera.url_history &> /dev/null
	echo -e -n "$BLUE[$GREEN*$BLUE] Cache cleaned\n"
	notify "Cache cleaned"
}


function ip {

	MYIP=`wget -qO- https://start.parrotsec.org/ip/`
	echo -e "\nMy ip is:\n"
	echo $MYIP
	echo -e "\n"
	notify "My IP is:\n\n$MYIP"
}


function start {
	# Make sure only root can run this script
	ME=$(whoami | tr [:lower:] [:upper:])
	if [ $(id -u) -ne 0 ]; then
		echo -e -e "\n$GREEN[$RED!$GREEN] $RED $ME R U DRUNK?? This script must be run as root$RESETCOLOR\n" >&2
		exit 1
	fi

	echo -e "\n$GREEN[$BLUE i$GREEN ]$BLUE Starting anonymous mode:$RESETCOLOR\n"

	if [ ! -e /tmp/tor.pid ]; then
		echo -e " $RED*$BLUE Tor is not running! $GREEN starting it $BLUE for you" >&2
		echo -e -n "\n $GREEN*$BLUE Stopping service nscd"
		/usr/sbin/service nscd stop 2>/dev/null || echo " (already stopped)"
		echo -e -n "\n $GREEN*$BLUE Stopping service resolvconf"
		/usr/sbin/service resolvconf stop 2>/dev/null || echo " (already stopped)"
		echo -e -n "\n $GREEN*$BLUE Stopping service dnsmasq"
		/usr/sbin/service dnsmasq stop 2>/dev/null || echo " (already stopped)"
		killall dnsmasq nscd resolvconf 2>/dev/null || true
		sleep 2
		killall -9 dnsmasq 2>/dev/null || true
		systemctl start tor
		sleep 20
	fi


	if ! [ -f /etc/network/iptables.rules ]; then
		/usr/sbin/iptables-save > /etc/network/iptables.rules
		echo -e "\n $GREEN*$BLUE Saved iptables rules\n"
	fi

	/usr/sbin/iptables -F
	/usr/sbin/iptables -t nat -F

	mv /etc/resolv.conf /etc/resolv.conf.bak
	echo -e 'nameserver 127.0.0.1' > /etc/resolv.conf
	echo -e " $GREEN*$BLUE Modified resolv.conf to use Tor and ParrotDNS/OpenNIC\n"

	# disable ipv6
	echo -e " $GREEN*$BLUE Disabling IPv6 for security reasons\n"
	/sbin/sysctl -w net.ipv6.conf.all.disable_ipv6=1
	/sbin/sysctl -w net.ipv6.conf.default.disable_ipv6=1

	# set iptables nat
	echo -e " $GREEN*$BLUE Configuring iptables rules to route all traffic through tor\n"
	/usr/sbin/iptables -t nat -A OUTPUT -m owner --uid-owner $TOR_UID -j RETURN

	#set dns redirect
	echo -e " $GREEN*$BLUE Redirecting DNS traffic through tor\n"
	/usr/sbin/iptables -t nat -A OUTPUT -p udp --dport 53 -j REDIRECT --to-ports 53
	/usr/sbin/iptables -t nat -A OUTPUT -p tcp --dport 53 -j REDIRECT --to-ports 53
	/usr/sbin/iptables -t nat -A OUTPUT -p udp -m owner --uid-owner $TOR_UID -m udp --dport 53 -j REDIRECT --to-ports 53

	#resolve .onion domains mapping 10.192.0.0/10 address space
	/usr/sbin/iptables -t nat -A OUTPUT -p tcp -d 10.192.0.0/10 -j REDIRECT --to-ports $TOR_PORT
	/usr/sbin/iptables -t nat -A OUTPUT -p udp -d 10.192.0.0/10 -j REDIRECT --to-ports $TOR_PORT

	#exclude local addresses
	for NET in $TOR_EXCLUDE 127.0.0.0/9 127.128.0.0/10; do
		/usr/sbin/iptables -t nat -A OUTPUT -d $NET -j RETURN
		/usr/sbin/iptables -A OUTPUT -d "$NET" -j ACCEPT
	done

	#redirect all other output through TOR
	/usr/sbin/iptables -t nat -A OUTPUT -p tcp --syn -j REDIRECT --to-ports $TOR_PORT
	/usr/sbin/iptables -t nat -A OUTPUT -p udp -j REDIRECT --to-ports $TOR_PORT
	/usr/sbin/iptables -t nat -A OUTPUT -p icmp -j REDIRECT --to-ports $TOR_PORT

	#accept already established connections
	/usr/sbin/iptables -A OUTPUT -m state --state ESTABLISHED,RELATED -j ACCEPT

	#allow only tor output
	echo -e " $GREEN*$BLUE Allowing only tor to browse in clearnet\n"
	/usr/sbin/iptables -A OUTPUT -m owner --uid-owner $TOR_UID -j ACCEPT
	/usr/sbin/iptables -A OUTPUT -j REJECT

	echo -e "$GREEN *$BLUE All traffic was redirected throught Tor\n"
	echo -e "$GREEN[$BLUE i$GREEN ]$BLUE You are under AnonSurf tunnel$RESETCOLOR\n"
	notify "Global Anonymous Proxy Activated"
	sleep 1
	notify "Dance like no one's watching. Encrypt like everyone is :)"
	sleep 10
}


function stop {
	# Make sure only root can run our script
	ME=$(whoami | tr [:lower:] [:upper:])

	if [ $(id -u) -ne 0 ]; then
		echo -e "\n$GREEN[$RED!$GREEN] $RED $ME R U DRUNK?? This script must be run as root$RESETCOLOR\n" >&2
		exit 1
	fi

	echo -e "\n$GREEN[$BLUE i$GREEN ]$BLUE Stopping anonymous mode:$RESETCOLOR\n"

	/usr/sbin/iptables -F
	/usr/sbin/iptables -t nat -F
	echo -e "\n $GREEN*$BLUE Deleted all iptables rules"

	if [ -f /etc/network/iptables.rules ]; then
		/usr/sbin/iptables-restore < /etc/network/iptables.rules
		rm /etc/network/iptables.rules
		echo -e "\n $GREEN*$BLUE Iptables rules restored"
	fi
	echo -e -n "\n $GREEN*$BLUE Restore DNS service"
	rm /etc/resolv.conf || true
	ln -s /etc/resolvconf/run/resolv.conf /etc/resolv.conf || true

	# re-enable ipv6
	/sbin/sysctl -w net.ipv6.conf.all.disable_ipv6=0
	/sbin/sysctl -w net.ipv6.conf.default.disable_ipv6=0

	/usr/sbin/service tor stop
	sleep 2
	killall tor
	sleep 6
	echo -e -n "\ $GREEN*$BLUE Restarting services\n"
	/usr/sbin/service resolvconf start || service resolvconf restart || true
	/usr/sbin/service dnsmasq start || true
	/usr/sbin/service nscd start || true
	echo -e " $GREEN*$BLUE It is safe to not worry for dnsmasq and nscd start errors if they are not installed or started already."
	sleep 1

	echo -e " $GREEN*$BLUE Anonymous mode stopped\n"
	notify "Global Anonymous Proxy Closed - Stop dancing :("
	sleep 4
}


function change {
	exitnode-selector
	sleep 10
	echo -e " $GREEN*$BLUE Tor daemon reloaded and forced to change nodes\n"
	notify "Identity changed - let's dance again!"
	sleep 1
}


function status {
	/usr/sbin/service tor@default status
	cat /tmp/anonsurf-tor.log || cat /var/log/tor/log
}



case "$1" in
	start)
		zenity --question --text="Do you want anonsurf to kill dangerous applications and clean some application caches?" --width 400 && init
		start
	;;
	stop)
		zenity --question --text="Do you want anonsurf to kill dangerous applications and clean some application caches?" --width 400 && init
		stop
	;;
	changeid|change-id|change)
		change
	;;
	status)
		status
	;;
	myip|ip)
		ip
	;;
	mac|mymac)
		mac
	;;
	restart)
		$0 stop
		sleep 1
		$0 start
	;;
   *)
echo -e "
Parrot AnonSurf Module (v 2.9)
	Developed by Lorenzo \"Palinuro\" Faletra <palinuro@parrotsec.org>
		     Lisetta \"Sheireen\" Ferrero <sheireen@parrotsec.org>
		     Francesco \"Mibofra\" Bonanno <mibofra@parrotsec.org>
		and a huge amount of Caffeine + some GNU/GPL v3 stuff
	Extended by Daniel \"Sawyer\" Garcia <dagaba13@gmail.com>

	Usage:
	$RED┌──[$GREEN$USER$YELLOW@$BLUE`hostname`$RED]─[$GREEN$PWD$RED]
	$RED└──╼ \$$GREEN"" anonsurf $RED{$GREEN""start$RED|$GREEN""stop$RED|$GREEN""restart$RED|$GREEN""change$RED""$RED|$GREEN""status$RED""}

	$RED start$BLUE -$GREEN Start system-wide TOR tunnel
	$RED stop$BLUE -$GREEN Stop anonsurf and return to clearnet
	$RED restart$BLUE -$GREEN Combines \"stop\" and \"start\" options
	$RED changeid$BLUE -$GREEN Restart TOR to change identity
	$RED status$BLUE -$GREEN Check if AnonSurf is working properly
	$RED myip$BLUE -$GREEN Check your ip and verify your tor connection
	$RED mymac$BLUE -$GREEN Check your mac and verify your change mac address
$RESETCOLOR
Dance like no one's watching. Encrypt like everyone is.
" >&2

exit 1
;;
esac

echo -e $RESETCOLOR
exit 0
